### ISSUE TITLE
Add custom resource scheduling

### ISSUE THREAD

**Issue by kewenwang-ai** [kubernetes/kubernetes#82118]
**What would you like to be added**:
Add custom resource scheduling, details as follows:
1. Besides the current scheduling of pod, Kube-Scheduler should support for scheduling CRD resources, which also as a scheduling unit.
2. Kube-Scheduler call the user-defined CRD 'split' policy via hooks. The 'split' policy is used to split the CRD into pods.
3. The pods split from one CRD object should be scheduled by atom, which means all or nothing. Scheduler cannot schedule other pod in the process of CRD scheduling. (Same as [Coscheduling/Gang-scheduling](https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/34-20180703-coscheduling.md))
4. Kube-Scheduler determines which CRDs to support via configuration.
5. Also support CRD priority-based scheduling, same as pod priority policy.

**Why is this needed**:
In some cases, such as deep learning and big data jobs should be scheduled as a whole. In kubernetes, we can use CRD to define the above job objects. But as the scale of job increases, cross-node distribution is an inevitable solution, so kubernetes should use multiple pods to host the workers for the job.

However, Kube-Scheduler only supports pod scheduling, either the pod is pre-defined by the user, or is defined by the user-defined CRD controller. Because the cluster resource usage cannot be obtained in real time, the pod that defined by the above two methods may has the following problems and irrationality:

1. Resource contention deadlock: Pods after the split of CRD resources cannot be atomically scheduled. As a result, when some pods of a CRD resource are scheduled, the remaining pods cannot be scheduled, and the currently occupied resources cannot be released, which may cause resource deadlock.
2. Non-optimal resource allocation: The user or CRD controller cannot know the current usage of cluster resources, which leads to non-optimal utilization of the resources when dividing pods.
3. Hunger problem: If the available resources of the cluster cannot satisfy the large-grained job request, the job will not be scheduled for a long time especially when small-grained jobs or pods are inserted.

Kubernetes need parallel scheduling because of deep learning jobs and other applications may coexist in the same Kubernetes cluster. [Multiple schedulers](https://kubernetes.io/docs/tasks/administer-cluster/configure-multiple-schedulers/) is a solution, however, it may introduce resource scheduling conflicts. So we propose this issue.

Here are some examples to illustrate the problems:

**example 1: (deadlock)**
There is a cluster with 4 available GPUs, a CRD A and a CRD B each apply for 3 GPUs, and both A and B are split into 3 pods, one GPU per pod. Then the following situation may occur when scheduling without atomic:
CrdA-Pod1 —> CrdA-Pod2 —> CrdB-Pod1 —> CrdB-Pod2 —> both A and B wait for resource release, deadlock

**example 2: (non-optimal)**
The following figure shows the current resource usage of a cluster. The user wants to start a deep learning job through 8 GPUs and define two 4 GPU pods. Although the two 4GPU Pods can be successfully scheduled, it is not the optimal allocation scheme. Because if the training job is allocated on one 8 GPU Pod, the interaction between Pod (Node) can be avoided, and the training efficiency will be higher.
![issue_2](https://user-images.githubusercontent.com/54662167/63923554-53f61300-ca79-11e9-8395-70821469e0e8.png)

**example 3: (hunger)**
The following figure shows the current resource usage of a cluster. When a new deep learning job request 8 GPUs and define one Pod (Worker) with 8 GPUs, although the total amount of remaining resources in the cluster is greater than 8, the job cannot be scheduled because no node satisfies the Pod's resource request. At the same time, if other small-request jobs are inserted into the schedule, this job will not be scheduled for a long time.
![issue_1](https://user-images.githubusercontent.com/54662167/63923631-70924b00-ca79-11e9-9df9-8f4001e7bd49.png)

PS: 
1. Current scheduler extender only support pod scheduling, cannot be used for CRD resources.
2. Multi-Schedulers work in parallel may cause resource scheduling conflicts.

/sig scheduling

**Comment by tedyu** [comment]
For example 3, would priority-based scheduling help solve the problem ?
The job for one Pod (Worker) with 8 GPUs would need to have higher priority.

**Comment by zhangchengming** [comment]
> For example 3, would priority-based scheduling help solve the problem ?
> The job for one Pod (Worker) with 8 GPUs would need to have higher priority.

Priority-based schduling just sovle part of problems. If some small-grained pods of same or higher priority insert in scheduling, this pod will not get the chance to be scheduled.

**Comment by k82cn** [comment]
hm... I'm a little confuse on what we're going to do for this? The overall workflow in default scheduler is pod based, how that will support others? Do you mean we schedule CRD instead of Pods?

**Comment by kewenwang-ai** [comment]
> hm... I'm a little confuse on what we're going to do for this? The overall workflow in default scheduler is pod based, how that will support others? Do you mean we schedule CRD instead of Pods?

Hi，thanks for your attention. We don't want to use CRD instead of pod scheduling. After our team's previous analysis, our ideas are as follows:

1. Let kube-scheduler support CRD scheduling, but does not change the original process and strategy of the pod scheduling.
2. However, scheduler doesn't need to understand CRD itself.
3. Use a scheduler extender(web hook) to complete the function that split crd to  pods.

**Comment by Adirio** [comment]
I think this could be solved if the scheduler was modified to be able to schedule a list of schedulers per cycle instead of only 1 and the list was passed as a pointer to the PreFilter plugins. A PreFilter plugin could append to this slice all the pods that are to be atomically scheduled. For example, a stateful PreFilter plugin could start caching which pods need to be scheduled, returning an error when some of them are still missing and returning the full list when the last one arrives.

**Comment by tedyu** [comment]
How would the potential conflict between PreFilter plugins be tracked ?

e.g. PreFilter plugins X and Y may have contradictory decision about pods A and B - whether they should be scheduled and the order.

**Comment by Adirio** [comment]
I think that is part of the work that a PreFilter should do. A PreFilter could order the pods, if they should be scheduled, etc. If there are multiple PreFilters that take this decissions, their order and configuration would determine the result, I guess.

**Comment by kewenwang-ai** [comment]
Hi, I guess you mean scheduler can schedule a list of pods per cycle?  and the PreFilter will wait all the pods belonging to one CRD and then return the pod list to next step Filter for atomically scheduling. am I understand right? @Adirio 
This idea is a little bit like [co-scheduling in the Permit plugin](https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/20180409-scheduling-framework.md#coscheduling). And also we could order the related pods in the Sort plugin. This method could slove the atomic scheduling.

But there is another question before atomic scheduling: how and when to split one CRD into pods? Because we hope to split the CRD according to the current cluster resource state, that's why we not choose controller or user-define(like [kubeflow/tf-job](https://www.kubeflow.org/docs/components/training/tftraining/)).

So we propose a Split interface of extender. And this could also be implemented by framework, for example, add a 'Split' extension point before Pre-Filter.
But anyway, currently we only think of methd that the Scheduler Queue should support both CRD and pod to slove this issue.
@tedyu

**Comment by Adirio** [comment]
You udenrstood correctly.

The difference between the co-scheduling in the Permit plugin section that you mentioned is that pods that are waiting in the permit hook have already a node assigned while if they wait in the PreFilter phase and then can be handled as a slice of pods instead of a single pod the decission of where to place each of them can depend on the rest, as you still haven't filtered nor scored each node for each pod.

About the splitting problem. Why do you think that splitting at the PreFilter hook would be different from splitting with a controller? The cluster can still change from the PreFilter to the Binding phase. I think that adding support for totally generic CRD resources instead of only pods in the scheduler will not improve your situation and is a really hard task while scheduling a list of pods atomically would allow you to solve your problem the same way and is much easier.

**Comment by alculquicondor** [comment]
If I understood correctly, these are the requirements you have:

1. You want to schedule a set of pods atomically.
1. These pods require GPUs ([device plugins](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/) in k8s vocabulary)
1. You want the pods to run in the same node, if there is enough space.

- 1 should be solved by co-scheduling, either via [kube-batch](https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/34-20180703-coscheduling.md) or with a future [Co-scheduling permit plugin](https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/20180409-scheduling-framework.md#coscheduling).
- 2: Do we already support this?
- 3 can be solved via [pod affinity](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity) `preferredDuringSchedulingIgnoredDuringExecution`.

Did I get your requirements right?
Rather than implementing another feature, would a mix of (existing) Pod Affinity and (future) Co-scheduling solve the problem?

Otherwise, can you clarify what would be the input and output of the splitting interface that you mentioned?

**Comment by kewenwang-ai** [comment]
> If I understood correctly, these are the requirements you have:
> 
> 1. You want to schedule a set of pods atomically.
> 2. These pods require GPUs ([device plugins](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/) in k8s vocabulary)
> 3. You want the pods to run in the same node, if there is enough space.
> 
> * 1 should be solved by co-scheduling, either via [kube-batch](https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/34-20180703-coscheduling.md) or with a future [Co-scheduling permit plugin](https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/20180409-scheduling-framework.md#coscheduling).
> * 2: Do we already support this?
> * 3 can be solved via [pod affinity](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity) `preferredDuringSchedulingIgnoredDuringExecution`.
> 
> Did I get your requirements right?
> Rather than implementing another feature, would a mix of (existing) Pod Affinity and (future) Co-scheduling solve the problem?
> 
> Otherwise, can you clarify what would be the input and output of the splitting interface that you mentioned?

Hi, thanks for responding so soon.
the third "You want the pods to run in the same node" is not correct.
What we want is the pods number for the crd to be  optimal, because 2 pods with 4GPUs is more efficient than 4 pod with 2GPUs. they don't have to be in the same node, they could run distributed.
Beacuse only the scheulder know the current resource usage of cluster, so scheduler could make the optimal decision on how many pods run for the job. 

yes, the kube-batch could solve the  atomic scheduling problem, but cannot solve the resource efficiency.

**Comment by ahg-g** [comment]
> But there is another question before atomic scheduling: how and when to split one CRD into pods? Because we hope to split the CRD according to the current cluster resource state, that's why we not choose controller or user-define(like [kubeflow/tf-job](https://www.kubeflow.org/docs/components/training/tftraining/)).

You can build a custom controller for that. The node objects contain the resources available on each node and how much is allocated, the controller would watch for node objects to determine how much resources available in the cluster and based on that split the "CRD" into pods.

**Comment by Adirio** [comment]
Adding to @ahg-g, the custom controller could start by using the prefered configuration that is the most strict (2 pods with 4 GPUs), and if it sees that the pods are failing relax the requirements (4 pods with 2 GPUs), or even a mix if one of them was able to start (1 pod with 4 GPUs, 2 pods with 2 GPUs).

**Comment by kewenwang-ai** [comment]
> > But there is another question before atomic scheduling: how and when to split one CRD into pods? Because we hope to split the CRD according to the current cluster resource state, that's why we not choose controller or user-define(like [kubeflow/tf-job](https://www.kubeflow.org/docs/components/training/tftraining/)).
> 
> You can build a custom controller for that. The node objects contain the resources available on each node and how much is allocated, the controller would watch for node objects to determine how much resources available in the cluster and based on that split the "CRD" into pods.

Node objects available resource is actually the capacity that can be used for scheduling, not the left usable resources of node. The allocated resources is still unknown to the controller and only scheduler knows it.

This way is not real-time effective. Because when the scheduler cannot split 2 pods with 4GPUs in **t_1** time, doesn't mean it cannot satisfy 2 pods with 4GPUs in **t_2** time. Also there could be some time before the scheduler do the scheduling after the controller make the split decison, which means the status of cluster may be different. 
This also could not the optimal. @Adirio

**Comment by ahg-g** [comment]
> Node objects available resource is actually the capacity that can be used for scheduling, not the left usable resources of node. The allocated resources is still unknown to the controller and only scheduler knows it.

Right, but you can build that information by watching for pods. Each scheduled pod object contains the name of the node where it is scheduled, so you can add up the resources consumed by the pods scheduled on the same node to calculate the amount of used resources. That is how the schedule works, and the [daemonset controller](https://github.com/kubernetes/kubernetes/blob/26cc580e65b12d230c18dc9c947084dd918633a6/pkg/controller/daemon/daemon_controller.go#L1300) as well (which you can consider as a second scheduler).

**Comment by alculquicondor** [comment]
Please clarify 2 things:

- What is specified in the CRD? Make it as precise as possible so we can understand the use case
- What does it mean to schedule the CRD, before "splitting"? It looks like you want to "reserve" a set of resources in several nodes for a single CRD.

**Comment by kewenwang-ai** [comment]
> > Node objects available resource is actually the capacity that can be used for scheduling, not the left usable resources of node. The allocated resources is still unknown to the controller and only scheduler knows it.
> 
> Right, but you can build that information by watching for pods. Each scheduled pod object contains the name of the node where it is scheduled, so you can add up the resources consumed by the pods scheduled on the same node to calculate the amount of used resources. That is how the schedule works, and the [daemonset controller](https://github.com/kubernetes/kubernetes/blob/26cc580e65b12d230c18dc9c947084dd918633a6/pkg/controller/daemon/daemon_controller.go#L1300) as well (which you can consider as a second scheduler).

Yes, I understand. 

But there may exist a time lag between the scheduler and the controller.
After the controller split the crd into pod-group, there maybe a time lag before the scheduler do scheduling. And also there may be interspersed with other Pod scheduling or Pod resource released, which causes a change in resource state, so the pod-group maybe not the optimal when actually scheduling, even cannot be scheduled.
I describe the problem with a picture below.
![issue_reply1](https://user-images.githubusercontent.com/54662167/65475240-6d8c4e00-deb1-11e9-9a9e-4c0398589330.PNG)
which means the state of cluster in time_2 maybe already different with time_1.

**Comment by kewenwang-ai** [comment]
> Please clarify 2 things:
> 
> * What is specified in the CRD? Make it as precise as possible so we can understand the use case
> * What does it mean to schedule the CRD, before "splitting"? It looks like you want to "reserve" a set of resources in several nodes for a single CRD.

Here is an example of a CRD job (such as define the CRD as DLJob) for Deep Learning, which user only need to describe the total resources of the job and specify the container images.

```yaml
apiVersion: sample.com/v1
kind: DLJob
metadata:
  name: example-dl-job
spec:
  # all the pod type in this job, and the split interface will    
  # decide how many pod to run of each type according the current
  # resource status of cluster.
  images:
    worker: workerimage
    ps: psimage
  restartPolicy: failedPodRestart/resetJob/none
  priorityClassName: high-priority
  resource:  # describe the total resource for this job
    gpu: 8
    cpu: 8
    memory: 2000
  training_data_uri: /workspace/project1/dl_training_data  
  models_uri: /workspace/project1/job1/output_models/ 
```

This yaml declare a DLJob with the request for 8GPU and 8CPU, when do the splitting, this job could be split into [PodA, PodB] each with 4GPU or [PodA, PodB, PodC, PodD] each with 2GPU according to the current left resources of each node in the cluster. The implementation of split interface is through scheduler extender.



For question 2:

Scheduler schedule the CRD, taking the following steps:

1. first scheduler watch the crds through the dynamic informer and put the crd into the scheduling queue.(current the queue is only for pod, we hope to support crd.)
2. scheduler get item from the queue, if that is a crd, then scheduling it with two steps：
   - split this crd through scheduler extender interface, which return a pod-group(or list).
   - then scheduler schedule the pod-group atomically. 

The steps 2 above is an atomic scheduling. So don't need to reserve the resources, after the "splitting", scheduler will schedule the pod in the group one by one immediately. Because no other pod scheduling will be inserted in the process, so the state of cluster is consistent between the splitting and scheduleing.

**Comment by kewenwang-ai** [comment]
> Take into account that there may be other schedulers concurrently so what you call atomic scheduling (ensuring that the Pods are deployed in the state that is viewed by the spliter) can not be guaranteed. Additionally, Pods may fail in the process. A decission is only optimal in the context it is given, and when this context is ever changing as a Kubernetes cluster, making super complex algorithms that ensure optimal deployments in the current context may not be worth as the state will change and the optimal decission you took may not be the optimal decission in the new state.
> 
> The points I want to make are:
> 
> 1. Even by doing the splitting in the scheduler, atomicity can _not_ be guaranteed due to other state changes coming from ending/failing pods or concurrent schedulers.
> 2. Even if the atomicity could be guaranteed, the state of the cluster will probably change during the execution making the previously considered optimal deployment non-optimal.

I understand. But that's also why we don't want to use concurrent schedulers(multi-schedulers) from the beginning, because multi-schedulers could bring resource conflicts. So we hope all the schuduling should be done in one scheduler, and the atomicity could be guaranteed.

Even original pod scheduling is based on the snapshot of current node state. It is optimal at the scheduling time.

**Comment by Adirio** [comment]
The scheduler cannot guarantee that there are no other schedulers, you would have to force that yourself. Even if you ensure that there is no second scheduler, there are still other sources that can modify the cluster state between the snapshot is taken until the decission is made (e.g., controller daemonset, failing or finishing pods).
As such, you will never be able to guarantee atomicity. And atomicity is an absolute measurement, getting close to atomic scheduling (without getting to fully atomic operations) provides no value. You will definetely get shorter times from the snapshot to the final decission if you modify the scheduler instead of going for the suggested controller approach, but that time will still be there.

I don't see this being generic enough to be included in the general scheduler. If you are worried about this time being too long if you go for the controller approach, why don't you make the scheduling decission in the controller the same way the daemonset controller decides where each demon is going?

**Comment by alculquicondor** [comment]
Let me state the steps (almost requirements) in more low level terms that I can extract from this request:

1. Make it possible to add CRDs to the scheduling queue.
1. Schedule the CRD by "reserving" resources in more than one node.
1. Call an external API to transform the CRD into pods. The call would contain a list of (node, reserved resources).
1. Schedule the pods in the same cycle.

Note that "reserve" is referring to an intermediate step within one scheduling cycle.
Do you think this would solve your use case? I would say that we are talking about very intrusive changes to the current scheduler, though.

**Comment by kewenwang-ai** [comment]
> Let me state the steps (almost requirements) in more low level terms that I can extract from this request:
> 
> 1. Make it possible to add CRDs to the scheduling queue.
> 2. Schedule the CRD by "reserving" resources in more than one node.
> 3. Call an external API to transform the CRD into pods. The call would contain a list of (node, reserved resources).
> 4. Schedule the pods in the same cycle.
> 
> Note that "reserve" is referring to an intermediate step within one scheduling cycle.
> Do you think this would solve your use case? I would say that we are talking about very intrusive changes to the current scheduler, though.


Yes, the steps you understand  is right.
We realize that there is some intrusive of our requirements, and that's why we hope to discuss these requirements first.

We propose this requirement based on the following reasons:

- The solutions based on kube-batch([kube-flow](https://www.kubeflow.org/docs/about/kubeflow/)) are a little unfriendly to user. Because user don't know the current snapshot of scheduler, so it would be much more reasonable and better if the scheduler do this 'splitting'. Also it would be more friendly to user if they only need to request total resouces, not the details.
- At the same time, there is a growing requirement for deep learning on kubernetes, such as kube-flow mentioned above. We hope there could be a more general and effective soultion for these requirements.
- Finally, the design we proposed will not modify the default scheduling process of Pod, only an extension to it, the intrusion is very limited.

**Comment by kewenwang-ai** [comment]
> The scheduler cannot guarantee that there are no other schedulers, you would have to force that yourself. Even if you ensure that there is no second scheduler, there are still other sources that can modify the cluster state between the snapshot is taken until the decission is made (e.g., controller daemonset, failing or finishing pods).
> As such, you will never be able to guarantee atomicity. And atomicity is an absolute measurement, getting close to atomic scheduling (without getting to fully atomic operations) provides no value. You will definetely get shorter times from the snapshot to the final decission if you modify the scheduler instead of going for the suggested controller approach, but that time will still be there.
> 
> I don't see this being generic enough to be included in the general scheduler. If you are worried about this time being too long if you go for the controller approach, why don't you make the scheduling decission in the controller the same way the daemonset controller decides where each demon is going?

But if it is what you said, it seems there is no way to guarantee the atomicity, not even the current soultion. 

Of course you choose use one shceduler or more, but once you use mutiple schedulers, you have already introduced resource competition conflicts, no matter use our proposal or not.

We only hope a better soultion for this issue.

**Comment by alculquicondor** [comment]
Maybe it's time to get a more detailed design.
What I meant is that the changes are intrusive to the current scheduler design, it's input and output, not that Pod scheduling is affected.
In particular, there seem to be 2 requirements that are very hard to fulfill today:

1. Insert CRDs in the scheduling queue. Today, we only insert Pods, which have well define properties.
2. Reserve resources in more than one pod. Today, we only look at one node at a time (we take advantage of parallelism here).

So, we would like to know how to make these requirements fit within the current architecture and the scheduler framework.
Then, we need to weight the benefits of adding this complexity and compare it to the alternative:
Have a custom controller that schedules the pods based on your requirements. As @Adirio suggested, even moving the decisions to the main scheduler wouldn't prevent race conditions in a generic cluster.

Have you considered extending kube-batch instead?

**Comment by ahg-g** [comment]
To add to what @alculquicondor and @Adirio feedback, this change also requires scheduling pods that didn't go through the API server yet, and so all admission logic and webhooks will be bypassed.

**Comment by ahg-g** [comment]
> But if it is what you said, it seems there is no way to guarantee the atomicity, not even the current soultion.

Correct, k8s by design is eventually consistent: it does not guarantee that the cluster will always be in the expected state, but guarantees that eventually it will. 

For example, pods could get evicted by kubelet, and that will certainly break what is referred here as "atomic" scheduling. In summary, whether you implement the proposed solution or not, you will always need some sort of a controller that drives the state back to where you want it, that is not going to be the scheduler because the scheduler does not "reschedule" pods (take a look at the [descheduler](https://github.com/kubernetes-sigs/descheduler)).

**Comment by kewenwang-ai** [comment]
> 1. Insert CRDs in the scheduling queue. Today, we only insert Pods, which have well defined properties.

Yes, our design need to add the CRD into the scheduling queue.

Here is the architecture and components about our design, and the process is described in the figure below.
![architecture](https://user-images.githubusercontent.com/54662167/66464235-302edf80-eab1-11e9-94f0-5cd35acd384a.png)
Based on the figure above, we propose the following principles:
- Kube-Scheduler needs to support the scheduling of CRD, but it does not need to understand the CRD because of the diversity of CRD.
- The function that needs understand the CRD is implemented by the user according to the characteristics of the CRD, and the Scheduler calls this function through the Scheduler's Extender mechanism.

And based on this, the preliminary concept of the system is as follows:
- Kube-Scheduler: This part is a function extension to the default Scheduler, mainly including: 1)  Determine which CRDs need to be scheduled according to the configuration. 2) Implement the management of CRD and Pods in the unified scheduling queue. 3) Cooperate with Pod-Splitor to complete the atomic scheduling of CRD.
- Pod-Splitor: Its main function is response to the Scheduler's request. The Splitor split the CRD into Pods according to the system resource configuration and current resource usage, as well as the user-defined CRD distributed strategy. Then create these pods to API Server, and return the results to Scheduler.
- CRD-Controller: User-defined CRD lifecycle management system, such as state management, replica control, etc.

So, there are the benefits and complexity of our proposal.
- Benefits：The design will slove the problems(as we talked before) that controller or kube-batch may bring. At the same time, implement the atomic scheduling and improve the resource efficiency of cluster.
- Complexity：we need to extend the default scheduler as described above, which may be little intrusive to current scheduler.

We know about the kube-batch is focused on the atomic scheduling, mostly as a second scheduler. But anyway, we will think deeply about this suggestion.
@alculquicondor

**Comment by kewenwang-ai** [comment]
> To add to what @alculquicondor and @Adirio feedback, this change also requires scheduling pods that didn't go through the API server yet, and so all admission logic and webhooks will be bypassed.

About this question, because we need do the scheduling right after the 'splitting' interface, the creation of pod（which means go through the API server） could be done in the extender, if the creation failed, then the splitiing operation failed.
(Of course the scheduler should filter these pods to prevent them from entering into scheduling queue again.)

Yes, in order to keep the CRD in the expected state, we will always need the controller to guarantee that.
As we think now, when pod(belonging to one CRD) failed or be evicted by accidentally, the corresponding controller could be informed, then find the own CRD and recreate the CRD, or could simply create a new pod binding to this CRD.

Thanks for mention the descheuler, I think mainly about the Pod eviction and Affinity. Indeed it will affect the pod state if used, but we may extend some new policys to avoid this perhaps. We will consider about these situations.

**Comment by Adirio** [comment]
What you are proposing is to have a pseudo-CRD-controller (which you called Pod-Splitor) integrated inside the scheduling process.

### About the benefits:
- You claim that this will provide attomicity, which as I already told you, can't be guaranteed even with this setup as there are many other things that may interfere with the cluster state.
- You also claim that this will improve the resource efficiency, which is not technically correct. The proposed change itself (accept CRDs in the scheduling queue and handle them with an extender) doesn't improve resource efficiency. It would enable your use case which may increase the resource efficiency. *1

### About the drawbacks:
- You mention complexity (having to make all the queue CRD-agnostic, etc.) but you are missing a more important drawback. You are trying to execute a pseudo-CRD-controller during the most criticical orchestration process, the scheduling. Controllers benefit from concurrency but the scheduler doesn't, it is the current bottleneck. Introducing a controller would block it even further.

[1] I'm still not sure about this increase compared to a normal CRD-controller. The amount of interferences during the time between the normal CRD-controller and your proposed version would have to be measured. Clusters with few state changes will have very low probability that a relevant state change will happen during this time. By relevant I mean that it would affect the decission of the CRD-controller. In the other hand, a cluster with often state changes will have a very high probability, but the improvement will be valid only until some other relevant change happens, which will be soon as they happen often, and once this happens your enhancement has vanished.

**Comment by alculquicondor** [comment]
Additionally to the points made by @Adirio, a few more things to think about:

1. What is the data that the scheduler passes to the Pod-Splitor?
1. How does the design fit within the scheduling cycle? Should a CRD go through all the scheduling algorithms? Should the created pods go through all the algorithms too?

**Comment by kewenwang-ai** [comment]
I think the atomic scheduling here means that Pods of CRD are scheduled without inserting other Pods that do not belong to this CRD, not even a single Pod, just like the Kube-Batch and gang-sheduling. Thus ensuring that these Pods scheduled either all succeed or none.

I understand that what you mean by "no value" is that the cluster resource state may change during this process. For example Pods being released, however, all the resource scheduling is based on the current resource snapshot, and the current scheduler also does not consider the Pod being released when scheduling the Pod.

About the drawbacks you mentioned, introducing split process in the scheduling process may affect the scheduling efficiency, that's possible. However, similar problems also exist with the currently extender filter、priority interfaces.  we will consider a more efficient way to do it, such as framework.

@Adirio

**Comment by kewenwang-ai** [comment]
> Additionally to the points made by @Adirio, a few more things to think about:
> 
> 1. What is the data that the scheduler passes to the Pod-Splitor?
> 2. How does the design fit within the scheduling cycle? Should a CRD go through all the scheduling algorithms? Should the created pods go through all the algorithms too?

Q1 : Only the CRD info, the extender could keep local cache consistent with the scheduler.

Q2: CRD only need to be split into pods through the split interface. and the returned pods will go through all the algorithms.

**Comment by Adirio** [comment]
> I think the atomic scheduling here means that Pods of CRD are scheduled without inserting other Pods that do not belong to this CRD, not even a single Pod, just like the Kube-Batch and gang-sheduling. Thus ensuring that these Pods scheduled either all succeed or none.
>
> I understand that what you mean by "no value" is that the cluster resource state may change during this process. For example Pods being released, however, all the resource scheduling is based on the current resource snapshot, and the current scheduler also does not consider the Pod being released when scheduling the Pod.

We were deinetely not talking about the same "atomic" concept. By atomic you can refer to two different things:

1. Either all pods are shceduled or none is.
2. The cluster state is unable to change since the snapshot is taken until the decission is made.

As we already agreed the second point will never be achived. The first part will be easily implemented with the `Permit` api. A plugin could easily wait in the permit phase until all the pods that need to be scheduled atomically reach here, and `Allow` them when that happens, `Reject` them if somone doesn't reach this state or a timeout occurs.

> About the drawbacks you mentioned, introducing split process in the scheduling process may affect the scheduling efficiency, that's possible. However, similar problems also exist with the currently extender filter、priority interfaces. we will consider a more efficient way to do it, such as framework.

The filter and priority interface extenders are also blocking processes, yes, and thats one of the main reasons behind the scheduling framework you mentioned. The framework is already being implemented, adapting the queue to accept CRDs would be implemented over the framework so we should try to split this request into the different bits, which of them are solved by the framework in a easier manner and detect those that are not solved.

So lets compare using a normal controller to split pods and then feed them to the queue vs accepting the CRDs in the queue. As I mentioned, the atomicity (in the sense of either all are deployed or none is) of a set of pods will be easily implemented with a permit plugin, which doesn't require to add CRDs to the queue. So I only see one difference, the snapshot differences. The controller will use one snapshot to split, and each of the pods will use one snapshot to schedule themselves. None of those snapshots will be exact at scheduling time.

Accepting CRDs in the queue will be quite a hard process as it is fully customized for Pods. Does using a single (inexact) snapshot improve the performance compared to using multiple (inexact too) snapshots even in your case? Is this change generic enough to be deployed in kube-scheduler? Could we generalize this in any way that simplifies the implementation and make it generic enough to be implemented in kube-scheduler?

As part of your proposal, multiple pods would have to be scheduled in a single scheduling cycle. This also has its difficulties. An intermediate step could be changing the queue of pods to be a queue of lists of pods. If we add this concept of multiple pods in a single cycle the only difference from handling the atomicity (all-or-none) with a permit plugin (which will no longer be needed) is that they use the same snapshot. So the solutions would be:

1. CRD-Controller + Permit plugin: N+1 snapshots (one for splitting, one per pod), all-or-none atomicity.
2. CRD-Controller + List-aware scheduling queue: 2 snapshot (one for splitting, one for all pods), all-or-none atomicity.
3. CRD-aware scheduling queue: 1 snapshot, all-or-none atomicity.

**Comment by alculquicondor** [comment]
I would lean towards Option 1. The only changes to the scheduler would be a new gang-plugin, which we have already discussed before.
Although, I don't quite understand what you mean N+1 snapshots. Is this the number of different states at which a scheduling decision will be made?

**Comment by Adirio** [comment]
> Although, I don't quite understand what you mean N+1 snapshots. Is this the number of different states at which a scheduling decision will be made?

Yes, kind of. The number of different time instants when a split or scheduling decission will be made, as he was worried about this aspect, as they will have potentially different free resources in each time.

**Comment by alculquicondor** [comment]
If we have the permit plugin, and assuming that there are no other schedulers, the changes are more likely to be removed pods and, thus, more resources. Sure, the split might not be "optimal" anymore, but at least the pods should fit.

**Comment by Adirio** [comment]
There may be other pods scheduled in the mid time too.

**Comment by tangbo-sh** [comment]
Option 1 is not atomic scheduling(or gang-scheduling), essentially reserving resources until all Pods scheduling of CRD  succeeds or timeout fails. This may result in Pods of CRD scheduling failures or the split scheme is not optimal. So the issue problem is not resolved.

**Comment by Adirio** [comment]
Atomic scheduling is quite vague. It is atomic in the sense that either all of them are scheduled or none is. If none is the controller would make a different split decission based on the new cluster state and it will be tried again.

**Comment by kewenwang-ai** [comment]
Refer to the [gang-scheduling](https://docs.google.com/document/d/1AUwcvTtULNvow5M9e428FnlvINO1uQ7ojRoTGuTp4DA/edit#), we noticed that our proposal is similar with some ideas mentioned in this doc. Such as [QueueJob and scheduleBatch](https://docs.google.com/document/d/1AUwcvTtULNvow5M9e428FnlvINO1uQ7ojRoTGuTp4DA/edit#heading=h.ckn8nv2jj0xv).

What we proposed here is to use generic CRD instead of QueueJob and put the split policy into the scheduleBatch before scheduling pods. Although the implementation of the function maybe different, the process is the very similar, especially the queue to store both Pod and Job.

We hope to know why chose to put the Coscheduling/Gang-scheduling design in kube-batch at that time, and why not implement it in the default scheduler?  What are the main reasons and background ? @k82cn

**Comment by k82cn** [comment]
> We hope to know why chose to put the Coscheduling/Gang-scheduling design in kube-batch at that time, and why not implement it in the default scheduler? What are the main reasons and background ?

Gang-scheduling is only one feature of batch scheduler; for example,  [gang plugin](https://github.com/kubernetes-sigs/kube-batch/tree/master/pkg/scheduler/plugins/gang) is only ~100LOC in kube-batch. If we want to also support other features, therer'e several enhancements to default scheduler. We'd like to handle them together which may reduce refactor work later.

**Comment by kewenwang-ai** [comment]
What if we implement this proposal through the framework, not the extender?
1. It still seems that we need to add the generic-CRD into the scheduling queuing. We can sort the CRD and the pod according to its priority by 'Sort Plugin'.
2. Every cycle of framework, if we pick up a CRD from the scheduling queue, we could introduce a 'Split Plugin' to do the splitting. And schedule the pods through the scheduling cycle one by one.
3. The pod belong to one CRD should wait in the 'Permit Plugin' of bind cycle untill all the pods of this CRD scheduled succeeded(or a minAvailable number). 

We could schedule a CRD within a cache snapshot and also not affect the existing processes in this way. Is this more suitable? (we know still intrusive to the scheduling queue)
@alculquicondor @Adirio 

Also Kube-batch schedule minAvailable tasks of a job in one scheduling cycle in the 'allocate' action. Would these features be merged into the default scheduler in the future? (the implementation logic is very different with current scheduler)
@k82cn

**Comment by k82cn** [comment]
> Also Kube-batch schedule minAvailable tasks of a job in one scheduling cycle in the 'allocate' action. Would these features be merged into the default scheduler in the future?

It's hard to do that in a short time; kube-batch does scheduling in job/podgroup level, and default scheduler does scheduling in pod level; that makes them hard to merged. I'm working on the migration, but I don-t think we can do that in a short time :)

**Comment by Adirio** [comment]
The third point is doable, as we already said. The first (and by extension the second) is really intrusive. Aside from all of the arguments against having CRDs in the scheduling queue we already gave, there is one we haven't consider: resource ownership.

Simplifying things, deployments are owned by the deployment controller. When a change is made to its spec, the deployment controller will handle it and knows nobody is going to modify the deployment status except himself. ReplicSets are owned by the replica set controller. And Pods are owned by the scheduler in first instance and then by the node where it was deployed.

If we introduce generic resources into the scheduling queue, that resource would be shared by its controller and the scheduler and that may be dangerous.

**Comment by kvaps** [comment]
So, any option to reuse a kube-scheduler for CRs nowadays?

**Comment by alculquicondor** [comment]
Maybe look at kueue.sigs.k8s.io

You can integrate a CRD like so: https://kueue.sigs.k8s.io/docs/tasks/integrate_a_custom_job/

**Comment by kvaps** [comment]
> Maybe look at https://kueue.sigs.k8s.io
> 
> You can integrate a CRD like so: https://kueue.sigs.k8s.io/docs/tasks/integrate_a_custom_job/

Hey @alculquicondor thanks for the tip, but unfortunately this not works for me, as my resources are represent not workload but storage replicas.

I want to schedule DRBD resources the same way like Kubernetes do for pods, with `nodeAffinity` and `topologySpreadConstraints` rules, but no pods are going to be run, instead these resources will be proceeded by specific controller on the nodes.

**Comment by alculquicondor** [comment]
Can you come to a sig-scheduling meeting to clarify your use case?

As it is, it's not clear to me why you wouldn't associate compute to these resources, in which case you could use a StatefulSet.

Or how generic this problem is so that kube-scheduler needs to implement it, as opposed to you building your own operator.
